modules/light-clients/08-wasm/light_client_module_test.go

```diff
@@ -77,209 +186,1378 @@ func (suite *WasmTestSuite) TestRecoverClient() {
+					var payload types.SudoMsg
+
+					err := json.Unmarshal(sudoMsg, &payload)
+					suite.Require().NoError(err)
+
+					expectedUpgradedClient, ok := upgradedClient.(*types.ClientState)
+					suite.Require().True(ok)
+					expectedUpgradedConsensus, ok := upgradedConsState.(*types.ConsensusState)
+					suite.Require().True(ok)
+
+					// verify payload values
+					suite.Require().Equal(expectedUpgradedClient.Data, payload.VerifyUpgradeAndUpdateState.UpgradeClientState)
+					suite.Require().Equal(expectedUpgradedConsensus.Data, payload.VerifyUpgradeAndUpdateState.UpgradeConsensusState)
+					suite.Require().Equal(upgradedClientProof, payload.VerifyUpgradeAndUpdateState.ProofUpgradeClient)
+					suite.Require().Equal(upgradedConsensusStateProof, payload.VerifyUpgradeAndUpdateState.ProofUpgradeConsensusState)
+
+					// verify other Sudo fields are nil
+					suite.Require().Nil(payload.UpdateState)
+					suite.Require().Nil(payload.UpdateStateOnMisbehaviour)
+					suite.Require().Nil(payload.VerifyMembership)
+					suite.Require().Nil(payload.VerifyNonMembership)
+
+					data, err := json.Marshal(types.EmptyResult{})
+					suite.Require().NoError(err)
+
+					// set new client state and consensus state
+					wrappedUpgradedClient := clienttypes.MustUnmarshalClientState(suite.chainA.App.AppCodec(), expectedUpgradedClient.Data).(*ibctm.ClientState)
+					store.Set(host.ClientStateKey(), clienttypes.MustMarshalClientState(suite.chainA.App.AppCodec(), upgradedClient))
+					store.Set(host.ConsensusStateKey(wrappedUpgradedClient.LatestHeight), clienttypes.MustMarshalConsensusState(suite.chainA.App.AppCodec(), upgradedConsState))
+
+					return &wasmvmtypes.ContractResult{Ok: &wasmvmtypes.Response{Data: data}}, wasmtesting.DefaultGasUsed, nil
+				})
+			},
+			nil,
+		},
+		{
+			"failure: invalid client state",
+			func() {
+				upgradedClient = &solomachine.ClientState{Sequence: 20}
+			},
+			clienttypes.ErrInvalidClient,
+		},
+		{
+			"failure: invalid height",
+			func() {
+				upgradedClient = &types.ClientState{LatestHeight: clienttypes.ZeroHeight()}
+			},
+			ibcerrors.ErrInvalidHeight,
+		},
+		{
+			"failure: cannot find client state",
+			func() {
+				clientID = unusedWasmClientID
+			},
+			clienttypes.ErrClientNotFound,
+		},
+		/* TODO(jim): Get back to these.
+		{
+			"failure: upgraded client state is not wasm client state",
+			func() {
+				// set upgraded client state to solomachine client state
+				upgradedClient = &types.ClientState{
+					Data:         nil,
+					Checksum:     nil,
+					LatestHeight: clienttypes.NewHeight(3, 0),
+				}
+			},
+			clienttypes.ErrInvalidClient,
+		},
+		{
+			"failure: upgraded consensus state is not wasm consensus state",
+			func() {
+				// set upgraded consensus state to solomachine consensus state
+				upgradedConsState = &solomachine.ConsensusState{}
+			},
+			clienttypes.ErrInvalidConsensus,
+		},
```

Comment:
Will you manage to fix this one? I tried something like `upgradedConsState = &solomachine.ConsensusState{Diversifier: "diversifier", Timestamp: 10000000}` but looks like the unmarshaling doesn't return an error, it just unmarshals to an empty 08-wasm consensus state...